examples/errorprone/StringFmtInPrintMethodsCheck.java

```diff
@@ -0,0 +1,112 @@
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.util.TreeScanner;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.tree.JCTree;
+import java.io.PrintStream;
+import java.util.List;
+import java.util.Objects;
+
+/** Matches on string formatting inside print methods. */
+@AutoService(BugChecker.class)
+@BugPattern(
+  name = "StringFmtInPrintMethodsCheck",
+  category = JDK,
+  summary = "String formatting inside print method",
+  severity = ERROR,
+  linkType = CUSTOM,
+  link = "github.com/google/startup-os/tree/master/examples/errorprone#StringFmtInPrintMethodsCheck"
+)
+public class StringFmtInPrintMethodsCheck extends BugChecker
+    implements MethodInvocationTreeMatcher {
+
+  Matcher<ExpressionTree> PRINT_METHOD =
+      instanceMethod().onDescendantOf(PrintStream.class.getName()).named("print");
+
+  Matcher<ExpressionTree> STRING_FORMAT =
+      staticMethod().onClass(String.class.getName()).named("format");
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!PRINT_METHOD.matches(tree, state)) {
+      return NO_MATCH;
+    }
+    Symbol base =
+        tree.getMethodSelect()
+            .accept(
+                new TreeScanner<Symbol, Void>() {
+                  @Override
+                  public Symbol visitIdentifier(IdentifierTree node, Void unused) {
+                    return ASTHelpers.getSymbol(node);
+                  }
+
+                  @Override
+                  public Symbol visitMemberSelect(MemberSelectTree node, Void unused) {
+                    return super.visitMemberSelect(node, null);
+                  }
+                },
+                null);
+    if (!Objects.equals(base, state.getSymtab().systemType.tsym)) {
+      return NO_MATCH;
+    }
+    ExpressionTree arg = Iterables.getOnlyElement(tree.getArguments());
+    if (!STRING_FORMAT.matches(arg, state)) {
+      return NO_MATCH;
+    }
+    List<? extends ExpressionTree> formatArgs = ((MethodInvocationTree) arg).getArguments();
+    return describeMatch(
+        tree,
+        SuggestedFix.builder()
+            .replace(
+                ((JCTree) tree).getStartPosition(),
+                ((JCTree) formatArgs.get(0)).getStartPosition(),
+                "System.err.printf(")
```

Comment:
Is it always `err` and not `out`?

Reply:
@oferb not necessarily -- but it's a _suggested_ fix, so user will be able to figure it out
If you remember, I vendored in this rule, not written 🙂

Reply:
Right.
Can we check if it was `err` or `out` and suggest accordingly?
Otherwise, this fix might end up being worse than before the fix, for this particular case, which is not so bad in the first place.

Reply:
@oferb fixed in 4d17c6f